File: jobboard/src/main/java/com/epitech/jobboard/Services/Job_applicationService.java
package com.epitech.jobboard.Services;

import java.util.List;

import javax.transaction.Transactional;

import com.epitech.jobboard.Entities.Job_application;

public interface Job_applicationService {
    /**
     * Creates a new Job_application
     * @param job_application A new Job_application
     */
    void create(Job_application job_application);

    /**
     * Updates the status of a job application.
     *
     * @param  job_application  the job application to update
     */
    @Transactional
    void updateStatus(Job_application job_application);

    /**
     * Deletes a Job_application
     * @param id The ID of the Job_application to delete
     */
    void deleteById(int id);
    
    /**
     * Retrieves a Job_application by its ID
     * @param id The ID of the Job_application
     * @return The Job_application with the specified ID
     */
    Job_application findById(int id);

    /**
     * Retrieves all Job_applications by Poste
     * @param id The ID of the Poste
     * @return A list of all Job_applications with the specified Poste
     */
    List<Job_application> findByPoste(int id);

    /**
     * Retrieves all Job_applications by Person
     * @param id The ID of the Person
     * @return A list of all Job_applications with the specified Person
     */
    List<Job_application> findByPerson(int id);

    /**
     * Retrieves all Job_applications
     * @return A list of all Job_applications
     */
    List<Job_application> findAll();
}

File: project-jobboard/src/Services/jobApplicationService.js
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:8080';

axios.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('userToken');
    if (token) {
      config.headers['Authorization'] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

/**
 * Retrieves all job applications from the API.
 *
 * @return {Promise} A promise that resolves to the response data.
 */
function getJobApplications() {
  return axios.get('/api/job_applications');
}

/**
 * Retrieves a job application by its ID.
 *
 * @param {number} id - The ID of the job application.
 * @return {Promise} A promise that resolves to the job application data.
 */
function getJobApplicationById(id) {
  return axios.get(`/api/job_applications/${id}`);
}

/**
 * Retrieves a job application by its poste ID.
 *
 * @param {number} id - The ID of the poste.
 * @return {Promise} A promise that resolves to the job application object.
 */
function getJobApplicationByIdPoste(id) {
  return axios.get(`/api/job_applications/poste/${id}`);
}

/**
 * Retrieves a job application by the ID of the person.
 *
 * @param {number} id - The ID of the person.
 * @return {Promise} A promise that resolves to the job application data.
 */
function getJobApplicationByIdPerson(id) {
  return axios.get(`/api/job_applications/person/${id}`);
}

/**
 * Creates a job application.
 *
 * @param {object} jobApplication - The job application to be created.
 * @param {number} jobApplication.advertisement_id - The ID of the advertisement.
 * @param {number} jobApplication.person_id - The ID of the person.
 * @param {boolean} jobApplication.email_sent - Indicates if email was sent.
 * @param {string} jobApplication.status - The status of the application.
 * @return {Promise} A promise that resolves to the created job application.
 */
function createJobApplication (jobApplication) {
  const data = {
    id: 0,
    advertissement: {
      id: jobApplication.advertisement_id
    },
    person: {
      id: jobApplication.person_id
    },
    email_sent: jobApplication.email_sent,
    status: jobApplication.status
  }
  return axios.post('/api/job_applications', data);
}

/**
 * Updates a job application.
 *
 * @param {object} jobApplication - The job application to be updated.
 * @param {number} jobApplication.advertisement_id - The ID of the advertisement.
 * @param {number} jobApplication.person_id - The ID of the person.
 * @param {boolean} jobApplication.email_sent - Indicates if email was sent.
 * @param {string} jobApplication.status - The status of the application.
 * @return {Promise} A promise that resolves to the updated job application.
 */
function updateJobApplication (jobApplication) {
  const data = {
    id: jobApplication.id,
    advertissement: {
      id: jobApplication.advertisement_id
    },
    person: {
      id: jobApplication.person_id
    },
    email_sent: jobApplication.email_sent,
    status: jobApplication.status
  }
  return axios.put(`/api/job_applications/`, data);
}

/**
 * Deletes a job application with the given ID.
 *
 * @param {number} id - The ID of the job application to delete.
 * @return {Promise} A promise that resolves to the result of the deletion.
 */
function deleteJobApplication(id) {
  return axios.delete(`/api/job_applications/${id}`);
}

const jobApplicationsService = {
  getJobApplications,
  getJobApplicationById,
  getJobApplicationByIdPerson,
  getJobApplicationByIdPoste,
  createJobApplication,
  updateJobApplication,
  deleteJobApplication
};

export default jobApplicationsService;
